pages/education.py

```diff
@@ -736,3 +810,155 @@ def render_education_sub_func(sub_func_data, country):
+                y=[(i + 1) / (n + 1) for i in range(n)]
+                + [(i + 1) / (n + 1) for i in range(n)],
+                x=[0.1 for i in range(n)] + [0.9 for i in range(n)],
+                color=node_colors,
+                customdata=node_custom_data,
+                hovertemplate="%{customdata[1]}:  %{customdata[0]}<extra></extra>",
+            ),
+            link=dict(
+                source=[i for i in range(data.shape[0])],
+                target=[data.shape[0] + dest.index(source[i]) for i in range(n)],
+                color=node_colors_opaque,
+                value=[1 for i in range(n)],
+                hovertemplate="Expenditure: %{source.customdata[0]} <br /> Attendance: %{target.customdata[0]}<extra></extra>",
+            ),
+        )
+    )
+
+    fig.add_annotation(
+        x=0.1,
+        y=1,
+        arrowcolor="rgba(0, 0, 0, 0)",
+        text=f"<b>Per Capita Expenditure on Education</b><br> <b>{base_year}</b>",
+    )
+    fig.add_annotation(
+        x=0.9,
+        y=1,
+        arrowcolor="rgba(0, 0, 0, 0)",
+        text=f"<b>Attendance</b> <br> <b>{base_year}</b>",
+    )
+
+    rank_mapping = {0: "1st", 10: "10th", 20: "20th", 30: "30th", 40: "40th"}
+    for i in range(0, n + 1, 10):
+        fig.add_annotation(
+            y=1 - ((i + 1) / (n + 1)),
+            x=0.075,
+            yshift=10,
+            text=f"<b>{rank_mapping[i]}</b>",
+            showarrow=False,
+        )
+
+    narrative = education_sub_narrative(base_year, data)
+    return fig, narrative
+
+
+def education_sub_narrative(year, data):
+
+    data["ROI"] = data.attendance / data.per_capita_expenditure
+    PCC = get_correlation_text(
+        data,
+        {"col_name": "ROI", "display": "ROI"},
+        {
+            "col_name": "per_capita_expenditure",
+            "display": "per capita expenditure on education",
+        },
+    )
+
+    narrative = f"In {year}, {PCC}"
+    best_ROI = data[data["ROI"] == data.ROI.max()].adm1_name.values[0]
+    worst_ROI = data[data["ROI"] == data.ROI.min()].adm1_name.values[0]
+
+    narrative += f" Among the subnational regions, in terms of return on public spending on education measured by attendance, {best_ROI} had the highest return on investment (ROI) while {worst_ROI} had the lowest."
+    return narrative
+
+
+@callback(
+    Output("year_slider_edu_container", "style"),
+    Output("year_slider_edu", "marks"),
+    Output("year_slider_edu", "value"),
+    Output("year_slider_edu", "min"),
+    Output("year_slider_edu", "max"),
+    Output("year_slider_edu", "tooltip"),
+    Input("stored-data-education-subnational", "data"),
+    Input("country-select", "value"),
+)
+def update_education_year_range(data, country):
+    try:
+        data = pd.DataFrame(data["edu_subnational_expenditure"])
+        data = data.loc[(data.func == "Education")]
+
+        data = filter_country_sort_year(data, country)
+        data = data[data["attendance"].notna()]
```

Comment:
Did you mean to have these be two datasets, one on exp and another on outcome and have respective years which then get passed to the `slider_helper` as args?

Reply:
Oh.. I think this part was half finished. Sorry!

I also forgot to include common years functionalities. (year data point shown in white if both outcome and expenditure are available and black if outcome data is not.) 

But I think we will have to revise this functionality when we want to make this slider as a global component. (adding new outcome visualizations, geoplot e.t.c) in the spatial tab. 
Let's leave it as it is for now.